amaranth/vendor/_altera.py

```diff
@@ -1,9 +1,161 @@
 from abc import abstractmethod
 
 from ..hdl import *
+from ..lib import io, wiring
 from ..build import *
 
 
+# The altiobuf_* and altddio_* primitives are explained in the following Intel documents:
+# * https://www.intel.com/content/dam/www/programmable/us/en/pdfs/literature/ug/ug_altiobuf.pdf
+# * https://www.intel.com/content/dam/www/programmable/us/en/pdfs/literature/ug/ug_altddio.pdf
+# See also errata mentioned in: https://www.intel.com/content/www/us/en/programmable/support/support-resources/knowledge-base/solutions/rd11192012_735.html.
+
+class InnerBuffer(wiring.Component):
+    """A private component used to implement ``lib.io`` buffers.
+
+    Works like ``lib.io.Buffer``, with the following differences:
+
+    - ``port.invert`` is ignored (handling the inversion is the outer buffer's responsibility)
+    - output enable is per-pin
+    """
+    def __init__(self, direction, port, *, useioff=False):
+        self.direction = direction
+        self.port = port
+        members = {}
+        if direction is not io.Direction.Output:
+            members["i"] = wiring.In(len(port))
+        if direction is not io.Direction.Input:
+            members["o"] = wiring.Out(len(port))
+            members["oe"] = wiring.Out(len(port))
+        super().__init__(wiring.Signature(members).flip())
+        if useioff:
+            if isinstance(port, io.SingleEndedPort):
+                port.io.attrs["useioff"] = 1
+            if isinstance(port, io.DifferentialPort):
+                port.p.attrs["useioff"] = 1
+                port.n.attrs["useioff"] = 1
+            if direction is not io.Direction.Input:
+                self.oe.attrs["useioff"] = 1
+
+    def elaborate(self, platform):
+        kwargs = dict(
+            p_enable_bus_hold="FALSE",
+            p_number_of_channels=len(self.port),
+        )
+        if isinstance(self.port, io.SingleEndedPort):
+            kwargs["p_use_differential_mode"] = "FALSE"
```

Comment:
You don't actually need to use this--[`Instance` accepts arguments of the form `("p", "blah", value)`](https://amaranth-lang.org/docs/amaranth/latest/guide.html#instances).